src/pages/api/auth/github/callback.ts

```diff
@@ -0,0 +1,93 @@
+import { APIRoute } from "astro";
+import {
+	fetchGitHubAccessToken,
+	fetchGitHubUser,
+} from "../../../../lib/game-saving/github";
+import {
+	getSession,
+	updateUserGitHubToken,
+} from "../../../../lib/game-saving/account";
+
+export const get: APIRoute = async ({ request, cookies }) => {
+	const url = new URL(request.url);
+	const code = url.searchParams.get("code");
+	const state = url.searchParams.get("state");
+
+	let userId = null;
+	try {
+		const stateObj = JSON.parse(decodeURIComponent(state as string));
+		userId = stateObj.userId;
+	} catch (error) {
+		console.error("State parsing error:", error);
+	}
+
+	if (!code || !userId) {
+		return new Response(
+			'<script>window.opener.postMessage({ status: "error", message: "Missing authorization code or invalid state" }, "*"); window.close();</script>',
+			{
```

Comment:
How do these script get executed?

Reply:
The script runs in the user's browser using window.opener.postMessage(), sending the auth status back to the originating window. After that, window.close() closes the popup or redirect window.

We listen for status and message on frontend and display or act depending on it :)